### ISSUE TITLE
Train and Val in the same directory

### ISSUE THREAD

**Issue by suzuki-sken** [ultralytics/yolov5#8773]
### Search before asking

- [X] I have searched the YOLOv5 [issues](https://github.com/ultralytics/yolov5/issues) and [discussions](https://github.com/ultralytics/yolov5/discussions) and found no similar questions.

### Question

I am very grateful for your help with YOLOv5.
I have a question about the training process.
In a previous document, I read a statement that Train and Val variables should be specified in the same directory.
If the same directory is specified for the Train and Val variables, will the training and validation data automatically be properly split by the data loader, etc.?

**Comment by glenn-jocher** [comment]
@suzuki-sken 👋 Hello! Thanks for asking about YOLOv5 🚀 **dataset formatting**. To train correctly your data must be in YOLOv5 format. If you specify the same path for train and val then the same data will be used for each, which is not recommended as you will be validating on your training data.

Please see our [Train Custom Data](https://docs.ultralytics.com/yolov5/tutorials/pytorch_hub_model_loading) tutorial for full documentation on dataset setup and all steps required to start training your first model. A few excerpts from the tutorial:

### 1.1 Create dataset.yaml

[COCO128](https://www.kaggle.com/datasets/ultralytics/coco128) is an example small tutorial dataset composed of the first 128 images in [COCO](http://cocodataset.org/#home) train2017. These same 128 images are used for both training and validation to verify our training pipeline is capable of overfitting. [data/coco128.yaml](https://github.com/ultralytics/yolov5/blob/master/data/coco128.yaml), shown below, is the dataset config file that defines 1) the dataset root directory `path` and relative paths to `train` / `val` / `test` image directories (or *.txt files with image paths), 2) the number of classes `nc` and 3) a list of class `names`:
```yaml
# Train/val/test sets as 1) dir: path/to/imgs, 2) file: path/to/imgs.txt, or 3) list: [path/to/imgs1, path/to/imgs2, ..]
path: ../datasets/coco128  # dataset root dir
train: images/train2017  # train images (relative to 'path') 128 images
val: images/train2017  # val images (relative to 'path') 128 images
test:  # test images (optional)

# Classes
nc: 80  # number of classes
names: [ 'person', 'bicycle', 'car', 'motorcycle', 'airplane', 'bus', 'train', 'truck', 'boat', 'traffic light',
         'fire hydrant', 'stop sign', 'parking meter', 'bench', 'bird', 'cat', 'dog', 'horse', 'sheep', 'cow',
         'elephant', 'bear', 'zebra', 'giraffe', 'backpack', 'umbrella', 'handbag', 'tie', 'suitcase', 'frisbee',
         'skis', 'snowboard', 'sports ball', 'kite', 'baseball bat', 'baseball glove', 'skateboard', 'surfboard',
         'tennis racket', 'bottle', 'wine glass', 'cup', 'fork', 'knife', 'spoon', 'bowl', 'banana', 'apple',
         'sandwich', 'orange', 'broccoli', 'carrot', 'hot dog', 'pizza', 'donut', 'cake', 'chair', 'couch',
         'potted plant', 'bed', 'dining table', 'toilet', 'tv', 'laptop', 'mouse', 'remote', 'keyboard', 'cell phone',
         'microwave', 'oven', 'toaster', 'sink', 'refrigerator', 'book', 'clock', 'vase', 'scissors', 'teddy bear',
         'hair drier', 'toothbrush' ]  # class names
```


### 1.2 Create Labels

After using a tool like [Roboflow Annotate](https://roboflow.com/annotate?ref=ultralytics) to label your images, export your labels to **YOLO format**, with one `*.txt` file per image (if no objects in image, no `*.txt` file is required). The `*.txt` file specifications are:

- One row per object
- Each row is `class x_center y_center width height` format.
- Box coordinates must be in **normalized xywh** format (from 0 - 1). If your boxes are in pixels, divide `x_center` and `width` by image width, and `y_center` and `height` by image height.
- Class numbers are zero-indexed (start from 0).

<img width="800" alt="Image Labels" src="https://user-images.githubusercontent.com/26833433/91506361-c7965000-e886-11ea-8291-c72b98c25eec.jpg">

The label file corresponding to the above image contains 2 persons (class `0`) and a tie (class `27`):

<p align="center"><img width="428" src="https://user-images.githubusercontent.com/26833433/112467037-d2568c00-8d66-11eb-8796-55402ac0d62f.png"></p>


### 1.3 Organize Directories

Organize your train and val images and labels according to the example below. YOLOv5 assumes  `/coco128` is inside a `/datasets` directory **next to** the `/yolov5` directory. **YOLOv5 locates labels automatically for each image** by replacing the last instance of `/images/` in each image path with `/labels/`. For example:
```bash
../datasets/coco128/images/im0.jpg  # image
../datasets/coco128/labels/im0.txt  # label
```

<p align="center"><img width="700" src="https://user-images.githubusercontent.com/26833433/134436012-65111ad1-9541-4853-81a6-f19a3468b75f.png"></p>
</details>


Good luck 🍀 and let us know if you have any other questions!

**Comment by suzuki-sken** [comment]
glenn-jocher, thank you for your prompt reply.

I understood that if I specify the same path for train and val, the same data will be used for each.

I was wondering if there is a method called autosplit in utils/dataloaders.py that implements a mechanism to automatically split the data.

Thanks for your help and further development of YOLOv5!

**Comment by glenn-jocher** [comment]
@suzuki-sken yes you can run the autosplit function manually on a dataset and then update your yaml accordingly.

**Comment by Irikos** [comment]
Hey @suzuki-sken, keep in mind that in the <your_dataset>.yaml you can specify the files that you want to train / val on, not necesarily the folders. You can even provide .txt files for train and val, which, in turn, contain all the images.

For example:
my_dataset.yaml
```
# Train/val/test sets as 1) dir: path/to/imgs, 2) file: path/to/imgs.txt, or 3) list: [path/to/imgs1, path/to/imgs2, ..]
path: ../datasets/training_data  # dataset root dir
train: ./training.txt  # train images (relative to 'path') 128 images
val: ./val.txt  # val images (relative to 'path') 128 images
test:  # test images (optional)
[...]
```

./training.txt
```
./images/1524.png
./images/1670.png
./images/724.png
./images/1692.png
./images/1629.png
./images/887.png
[...]
```


./val.txt
```
./images/572.png
./images/234.png
./images/1387.png
./images/299.png
./images/77.png
./images/1735.png
[...]

```

You can load all the file paths and apply `train_test_split()` on it and then write the results in test.txt and val.txt. Or, do a k-fold cross-validation on it, and again, write the results etc.

This will eliminate the need for the files (or labels) to be moved when splitting the dataset.

Keep in mind that the path in .txt files is relative to the dataset root dir path from yaml.

**Comment by suzuki-sken** [comment]
Thanks @Irikos for your reply.

I understand exactly how the autosplit function works and how to define the yaml file.

I would like to generate a yaml file using the autosplit function and verify the generalization performance with k-Fold Cross Validation.

Thanks again @Irikos and @glenn-jocher!

**Comment by gjr2019** [comment]
> 

Hi, I want to know that how can i do k-Fold Cross Validation on the data.yaml file

**Comment by Irikos** [comment]
hey @gjr2019, I can show you how I did it. There might be better ways, though.

Basically I loaded the entire files folder and the annotations folders, stored their paths into an array, did k-fold on that array and wrote the .yaml file, each time calling yolo on it.

Here's the actual code I used, but keep in mind that I had only one class, you will need to adapt it to multiple classes if that is the case.

'
X = np.array(read_images(DATASET_FOLDER))
y = read_annotations_files(ANNOTATIONS_FOLDER)

# print(X)
# print(y)
kf = KFold(n_splits=10)


DATASET_PATH = "../datasets/my_dataset/training_data"
TRAIN_FILE_PATH = "../datasets/my_dataset/training_data/training.txt"
VALIDATION_FILE_PATH = "../datasets/my_dataset/training_data/validation.txt"
DATASET_YAML_PATH = "../datasets/my_dataset/training_data/my_data.yaml"
TEST_FILE_PATH = ""
classes = ["my_class"]

i = 0

for train_index, validation_index in kf.split(X):
    print("Starting fold no: " + str(i))
    start_time = time.time()
    X_train, X_validation = X[train_index], X[validation_index]

    file = open(TRAIN_FILE_PATH, "w+")
    for item in X_train:
        file.write("./images/" + item + "\n")
    file.close()

    file = open(VALIDATION_FILE_PATH, "w+")
    for item in X_validation:
        file.write("./images/" + item + "\n")
    file.close()
    write_yolo_dataset_yaml(classes, DATASET_YAML_PATH, DATASET_PATH)
    
    current_iteration = "y5s_pre_K_" + str(i)
    
    !python ./train.py --img 640 --cfg ./models/yolov5s.yaml --hyp ./data/hyps/hyp.scratch-high.yaml --batch 128 --epochs 2000 --data ../datasets/my_dataset/training_data/my_data.yaml --weights "yolov5s.pt" --workers 24 --name $current_iteration
    
    e = int(time.time() - start_time)
    print('{:02d}:{:02d}:{:02d}'.format(e // 3600, (e % 3600 // 60), e % 60))
    print("Ended fold no: " + str(i) + "\n")
    i += 1
 
'


'### WRITE THE YOLO CUSTOM DATASET.YAML. Wrote with kfold validation in mind, but useful in general. ###
# INPUT:
#   yaml_path: dataset.yaml path
#   dataset_path: path to the dataset folder
#   train_file: path to file with training images paths (relative to dataset folder)
#   validation_file: path to file with validation images paths (relative to dataset folder)
#   test_file: path to file with test images paths (relative to dataset folder). Can be empty.
#   classes: array of class names to be written in the dataset (note: now classes are hardcoded)
def write_yolo_dataset_yaml(classes, yaml_path, dataset_path, train_file = "./training.txt", validation_file = './validation.txt', test_file = ''):
    file = open(yaml_path, "w+")

    file.write("path: " + dataset_path + "\n")
    file.write("train: " + train_file + "\n") # relative to dataset_path
    file.write("val: " + validation_file + "\n")
    file.write("test: " + test_file + "\n") # not used yet
    file.write("nc: " + str(len(classes)) + "\n")
    file.write("names: " + str(classes) + "\n")
    file.close()
'

**Comment by Irikos** [comment]
hey @gjr2019, I can show you how I did it. There might be better ways, though.

Basically I loaded the entire files folder and the annotations folders, stored their paths into an array, did k-fold on that array and wrote the .yaml file, each time calling yolo on it.

Here's the actual code I used, but keep in mind that I had only one class, you will need to adapt it to multiple classes if that is the case.

```

X = np.array(read_images(DATASET_FOLDER))
y = read_annotations_files(ANNOTATIONS_FOLDER)

# print(X)
# print(y)
kf = KFold(n_splits=10)


DATASET_PATH = "../datasets/my_dataset/training_data"
TRAIN_FILE_PATH = "../datasets/my_dataset/training_data/training.txt"
VALIDATION_FILE_PATH = "../datasets/my_dataset/training_data/validation.txt"
DATASET_YAML_PATH = "../datasets/my_dataset/training_data/my_data.yaml"
TEST_FILE_PATH = ""
classes = ["my_class"]

i = 0

for train_index, validation_index in kf.split(X):
    print("Starting fold no: " + str(i))
    start_time = time.time()
    X_train, X_validation = X[train_index], X[validation_index]

    file = open(TRAIN_FILE_PATH, "w+")
    for item in X_train:
        file.write("./images/" + item + "\n")
    file.close()

    file = open(VALIDATION_FILE_PATH, "w+")
    for item in X_validation:
        file.write("./images/" + item + "\n")
    file.close()
    write_yolo_dataset_yaml(classes, DATASET_YAML_PATH, DATASET_PATH)
    
    current_iteration = "y5s_pre_K_" + str(i)
    
    !python ./train.py --img 640 --cfg ./models/yolov5s.yaml --hyp ./data/hyps/hyp.scratch-high.yaml --batch 128 --epochs 2000 --data ../datasets/my_dataset/training_data/my_data.yaml --weights "yolov5s.pt" --workers 24 --name $current_iteration
    
    e = int(time.time() - start_time)
    print('{:02d}:{:02d}:{:02d}'.format(e // 3600, (e % 3600 // 60), e % 60))
    print("Ended fold no: " + str(i) + "\n")
    i += 1
```
 


```

### WRITE THE YOLO CUSTOM DATASET.YAML. Wrote with kfold validation in mind, but useful in general. ###
# INPUT:
#   yaml_path: dataset.yaml path
#   dataset_path: path to the dataset folder
#   train_file: path to file with training images paths (relative to dataset folder)
#   validation_file: path to file with validation images paths (relative to dataset folder)
#   test_file: path to file with test images paths (relative to dataset folder). Can be empty.
#   classes: array of class names to be written in the dataset (note: now classes are hardcoded)
def write_yolo_dataset_yaml(classes, yaml_path, dataset_path, train_file = "./training.txt", validation_file = './validation.txt', test_file = ''):
    file = open(yaml_path, "w+")

    file.write("path: " + dataset_path + "\n")
    file.write("train: " + train_file + "\n") # relative to dataset_path
    file.write("val: " + validation_file + "\n")
    file.write("test: " + test_file + "\n") # not used yet
    file.write("nc: " + str(len(classes)) + "\n")
    file.write("names: " + str(classes) + "\n")
    file.close()

```
